RavenDB-18201 - Sharding - Edit Document View Endpoints

lastav5 opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-18201

### Additional description

`GET /databases/*/revisions`
`GET /databases/*/revisions/count`
`GET /databases/*/timeseries/stats`

lastav5 committed:
RavenDB-18201 - Sharding - Edit Document View Endpoints - GET revisions

ppekrol commented:
test this please

lastav5 committed:
RavenDB-18201 - Sharding - Edit Document View Endpoints - revisions/count

karmeli87 commented on src/Raven.Server/Documents/Sharding/Handlers/Processors/Revisions/ShardedRevisionsHandlerProcessorForGetRevisions.cs:
```diff
@@ -0,0 +1,101 @@
+
+        protected override void CheckNotModified(string actualEtag)
+        {
+            DevelopmentHelper.ShardingToDo(DevelopmentHelper.TeamMember.Karmel, DevelopmentHelper.Severity.Minor, "Implement AddPagingPerformanceHint");
```
wrong message and please put Normal severity

lastav5 replied:
done

lastav5 committed:
RavenDB-18201 - Sharding - Edit Document View Endpoints - time series stats

lastav5 committed:
RavenDB-18201 - addressed PR comments

lastav5 committed:
RavenDB-18201 - Sharding - Edit Document View Endpoints - GET /databases/*/counters

lastav5 committed:
RavenDB-18201 - added tests for counters and time series

ppekrol commented on src/Raven.Client/Documents/Operations/TimeSeries/GetTimeSeriesStatisticsOperation.cs:
```diff
@@ -47,6 +47,12 @@ public override HttpRequestMessage CreateRequest(JsonOperationContext ctx, Serve
             {
+                if (response == null)
+                {
+                    Result = null;
```
How this can happen?

ppekrol commented on src/Raven.Server/Documents/Handlers/Processors/Counters/CountersHandlerProcessorForGetCounters.cs:
```diff
@@ -0,0 +1,41 @@
+        {
+        }
+
+        protected override async ValueTask<CountersDetail> GetCountersAsync(string docId, StringValues counters, bool full)
```
remove `await`

ppekrol commented on src/Raven.Server/Documents/Handlers/Processors/Revisions/RevisionsHandlerProcessorForGetRevisions.cs:
```diff
@@ -0,0 +1,173 @@
+                }
+
+                var actualEtag = ComputeHttpEtags.ComputeEtagForRevisions(revisions);
+                CheckNotModified(actualEtag);
```
this needs to brake if this is not modified

ppekrol commented on src/Raven.Server/Documents/Handlers/Processors/Revisions/RevisionsHandlerProcessorForGetRevisions.cs:
```diff
@@ -0,0 +1,173 @@
+                }
+
+                var actualChangeVector = revisions.Length == 0 ? "" : revisions[0].ChangeVector;
+                CheckNotModified(actualChangeVector);
```
this needs to brake if not modified. Probably you want to return bool from this method?

ppekrol replied:
Check how this is handled in `AbstractStudioCollectionsHandlerProcessorForPreviewCollection`, I would prefer to handle that in a similar way.

ppekrol commented on src/Raven.Server/Documents/Handlers/Processors/TimeSeries/TimeSeriesHandlerProcessorForGetTimeSeriesStats.cs:
```diff
@@ -0,0 +1,75 @@
+
+        protected override ValueTask<TimeSeriesStatistics> GetTimeSeriesStatsAsync(string docId)
+        {
+            var tsStats = new TimeSeriesStatistics()
```
this can be moved down, right?

ppekrol commented on src/Raven.Server/Documents/Handlers/Processors/Revisions/AbstractRevisionsHandlerProcessorForGetRevisions.cs:
```diff
@@ -0,0 +1,86 @@
+    public class RevisionsResult
+    {
+        public BlittableJsonReaderObject[] Results;
+        public long TotalResults;
```
Why this is needed?

ppekrol commented on src/Raven.Client/Json/BlittableArrayResult.cs:
```diff
@@ -5,6 +5,7 @@ namespace Raven.Client.Json
     public class BlittableArrayResult
     {
         public BlittableJsonReaderArray Results { get; set; }
+        public long TotalResults { get; set; }
```
Why this is needed?

ppekrol commented on src/Raven.Server/Documents/Sharding/Handlers/Processors/Revisions/ShardedRevisionsHandlerProcessorForGetRevisions.cs:
```diff
@@ -0,0 +1,101 @@
+            int shardIndex;
+            using (RequestHandler.ContextPool.AllocateOperationContext(out TransactionOperationContext context))
+            {
+                shardIndex = RequestHandler.DatabaseContext.GetShardNumber(context, id);
```
shardNumber

ppekrol commented on src/Raven.Server/Documents/Sharding/Handlers/Processors/Revisions/ShardedRevisionsHandlerProcessorForGetRevisions.cs:
```diff
@@ -0,0 +1,101 @@
+            var id = RequestHandler.GetQueryStringValueAndAssertIfSingleAndNotEmpty("id");
+            var before = RequestHandler.GetDateTimeQueryString("before", required: false);
+            var start = RequestHandler.GetStart();
+            var pageSize = RequestHandler.GetPageSize();
```
Something is not correct, I see a lot of code duplication between sharded and non-sharded version of this processor. Ping me tomorrow.

ppekrol commented on src/Raven.Server/Documents/Handlers/Processors/Revisions/RevisionsHandlerProcessorForGetRevisions.cs:
```diff
@@ -0,0 +1,173 @@
+                {
+                    var revisionResult = new RevisionsResult()
+                    {
+                        Results = revisions.Select(x => x?.Data).ToArray(),
```
Forgot about this one in my last review. Can we avoid LINQ?

lastav5 committed:
RavenDB-18201 - Sharding - Edit Document View Endpoints - DELETE /databases/*/attachments

lastav5 committed:
RavenDB-18201 - addressed PR comments and

lastav5 committed:
RavenDB-18201 - fixes after rebase

ppekrol committed:
RavenDB-18201 renames

ppekrol committed:
RavenDB-18201 cleanup

ppekrol committed:
RavenDB-18201
- use one context in AbstractCountersHandlerProcessorForGetCounters
- simply the abstraction

lastav5 committed:
RavenDB-18201 - debug github hangs

ppekrol committed:
RavenDB-18201 revamping SetupRevisions

ppekrol committed:
RavenDB-18201 implemented POST /databases/*/admin/revisions/config

ppekrol committed:
RavenDB-18201 fixing test

ppekrol committed:
RavenDB-18201 properly create operation tokens in ShardedDatabaseRequestHandler

ppekrol committed:
RavenDB-18201 fixing compilation

ppekrol committed:
RavenDB-18201 cleanup

ppekrol committed:
RavenDB-18201 more renames
